src/pcnt.rs

```diff
@@ -0,0 +1,543 @@
+    Disable,
+}
+
+impl Into<esp_idf_sys::pcnt_ctrl_mode_t> for PcntControlMode {
+    fn into(self) -> esp_idf_sys::pcnt_ctrl_mode_t {
+        match self {
+            PcntControlMode::Keep => esp_idf_sys::pcnt_channel_level_action_t_PCNT_CHANNEL_LEVEL_ACTION_KEEP,
+            PcntControlMode::Reverse => esp_idf_sys::pcnt_channel_level_action_t_PCNT_CHANNEL_LEVEL_ACTION_INVERSE,
+            PcntControlMode::Disable => esp_idf_sys::pcnt_channel_level_action_t_PCNT_CHANNEL_LEVEL_ACTION_HOLD,
+        }
+    }
+}
+
+bitflags! {
+    #[allow(dead_code)]
+    #[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
+    pub struct PcntEventType: u32 {
+        #[doc = "< PCNT watch point event: threshold1 value event"]
+        const THRES_1 = esp_idf_sys::pcnt_evt_type_t_PCNT_EVT_THRES_1;
+        #[doc = "< PCNT watch point event: threshold0 value event"]
+        const THRES_0 = esp_idf_sys::pcnt_evt_type_t_PCNT_EVT_THRES_0;
+        #[doc = "< PCNT watch point event: Minimum counter value"]
+        const L_LIM = esp_idf_sys::pcnt_evt_type_t_PCNT_EVT_L_LIM;
+        #[doc = "< PCNT watch point event: Maximum counter value"]
+        const H_LIM = esp_idf_sys::pcnt_evt_type_t_PCNT_EVT_H_LIM;
+        #[doc = "< PCNT watch point event: counter value zero event"]
+        const ZERO = esp_idf_sys::pcnt_evt_type_t_PCNT_EVT_ZERO;
+    }
+}
+
+#[doc = " @brief Pulse Counter configuration for a single channel"]
+pub struct PcntConfig<'d> {
+    #[doc = "< Pulse input GPIO number, if you want to use GPIO16, enter pulse_gpio_num = 16, a negative value will be ignored"]
+    pub pulse_pin: Option<&'d AnyInputPin>,
+    #[doc = "< Control signal input GPIO number, a negative value will be ignored"]
+    pub ctrl_pin: Option<&'d AnyInputPin>,
+    #[doc = "< PCNT low control mode"]
+    pub lctrl_mode: PcntControlMode,
+    #[doc = "< PCNT high control mode"]
+    pub hctrl_mode: PcntControlMode,
+    #[doc = "< PCNT positive edge count mode"]
+    pub pos_mode: PcntCountMode,
+    #[doc = "< PCNT negative edge count mode"]
+    pub neg_mode: PcntCountMode,
+    #[doc = "< Maximum counter value"]
+    pub counter_h_lim: i16,
+    #[doc = "< Minimum counter value"]
+    pub counter_l_lim: i16,
+    #[doc = "< the PCNT channel"]
+    pub channel: PcntChannel,
+}
+
+#[derive(Debug)]
+pub struct Pcnt {
+    unit: pcnt_unit_t,
+}
+
+impl<'d> Pcnt {
+    pub fn new() -> Result<Self, EspError> {
+        Ok(Pcnt {
+            unit: unit_allocate()?,
+        })
+    }
+
+    #[doc = " @brief Configure Pulse Counter unit"]
+    #[doc = "        @note"]
+    #[doc = "        This function will disable three events: PCNT_EVT_L_LIM, PCNT_EVT_H_LIM, PCNT_EVT_ZERO."]
+    #[doc = ""]
+    #[doc = " @param pconfig Reference of PcntConfig"]
+    #[doc = ""]
+    #[doc = " @return"]
+    #[doc = "     - ()"]
+    #[doc = "     - EspError"]
+    pub fn config(&mut self, pconfig: & PcntConfig) -> Result<(), EspError> {
```

Comment:
All other drivers take a `&config` instance as part of their `new` constructor. Please do this here as well. If your driver can be reconfigured, you might provide a `reconfigure` method.

Reply:
This configures a channel, not the unit itself.  However it may be thaty the high/low limit config values are actually a unit configuration as it it in idf 5 even tho its passed in the channel config here.

Reply:
Maybe rename this function as `channel_config`?  Or implement something closer to v5 where we call a method on the driver to allocate one of the channels?